### PR TITLE
configurable timeout for process starter

### PR DESCRIPTION
**Pull request by northernSage**
- [x] update process starter to take configurable timeout value
- [x] add test for timeout feature
- [x] add CHANGELOG.rst entry
- [x] write doc section documenting new feature in more detail

Addresses #51 

This is the simplest amongst multiple approaches I have considered, being them:

- Using [signal](https://docs.python.org/3/library/signal.html) to register a custom timeout handler for `xprocess.ensure` call upon `signal.SIGALRM`. Unfortunately, this is UNIX only due to the lack of [SIGALRM](https://docs.python.org/3/library/signal.html#signal.SIGALRM) on Windows
- Multithreading. Using a second thread to keep track of execution time, join threads and raise `TimeoutError` if call does not return within a given time. Here, there's risk of creating a hanging thread if `xprocess.ensure` call never returns. I would like to avoid having to abruptly kill threads and deal with potential problems related to it in the future.
- Multiprocess. This would probably work since a child processes can be used as described in the case of threads and just killed afterwards without complications. But there's the additional cost of having to spawn another process only to time a function call, we would also have to properly cleanup after it.

Reminding that (from #51):

> Currently xprocess.ensure will hang indefinitely in the event of an unmatched start pattern **given fewer lines than the number specified by ProcessStarter.max_read_lines have been printed**.

The problem happens when
```python
# ProcessStarter.get_lines
def get_lines(self, log_file):
    while True:
        line = log_file.readline()
        if not line:
            std.time.sleep(0.1) # <<< get stuck here
        yield line
```
runs out of lines to read before `max_read_lines` have been read and the pattern hasn't been matched yet.

I propose the following
```python
# ProcessStarter.get_lines
def get_lines(self, log_file):
    """Read and yield one line at a time from log_file. Will raise
    TimeoutError if a line read hangs for more than self.line_wait_timeout seconds"""
    step_counter = max_wait_steps = int(self.line_wait_timeout / 0.1)
    while True:
        line = log_file.readline()
        if not line:
            step_counter -= 1
            std.time.sleep(0.1)
            if step_counter <= 0:
                raise TimeoutError(
                    "The provided start pattern {} could not be matched \
                    within the specified time interval of {} seconds".format(
                    self.pattern, self.line_wait_timeout)
                )
        else:
            # reset counter
            step_counter = max_wait_steps
        yield line
```
the changes will make the timeout counter enter in action only after all available lines have been read (given pattern hasn't been matched yet and that line count is still under `max_lines_read` ). In other words, while the process is writing to `stdout/stderr ` and lines are being read by `ProcessStarter.get_lines` the timeout countdown will be disabled and will only kick in when the function hang on a line read (process stopped outputting), which is exactly when the problem happens.

### CODE REVIEW

File: xprocess.py [hunk 1]
```diff
@@ -203,37 +203,60 @@ class ProcessStarter:
...
+    line when trying to match pattern before raising TimeoutError"""
+
     def __init__(self, control_dir, process):
         self.control_dir = control_dir
         self.process = process
 
     @abc.abstractproperty
     def args(self):
         "The args to start the process"
+        raise NotImplementedError
```
Comment (nicoddemus) [inline xprocess.py:217]: Curious, why was this added? Doesn't the `abc.abstractproperty` decorator accomplishes the same?

File: xprocess.py [hunk 2]
```diff
@@ -203,37 +203,60 @@ class ProcessStarter:
...
         self.process.log.debug(line)
         return line
 
     def get_lines(self, log_file):
+        """Read and yield one line at a time from log_file. Will raise
+        TimeoutError if a line read hangs for more than self.line_wait_timeout
+        seconds"""
+        step_counter = max_wait_steps = int(self.line_wait_timeout / 0.1)
         while True:
+            print("** step counter: ", step_counter)
```
Comment (nicoddemus) [inline xprocess.py:245]: Left over?

File: xprocess.py [hunk 3]
```diff
@@ -203,37 +203,60 @@ class ProcessStarter:
...
+        """fetch first <max_read_lines>, ignoring blank lines"""
         non_empty_lines = (x for x in lines if x.strip())
         return itertools.islice(non_empty_lines, self.max_read_lines)
 
     def log_line(self, line):
+        """Write line to process log file"""
         self.process.log.debug(line)
         return line
 
     def get_lines(self, log_file):
```
Comment (nicoddemus) [inline xprocess.py:239]: This implementation looks good, but I would write it in terms of "elapsed" time rather than use a counter:

```python
    def get_lines(self, log_file):
        """Read and yield one line at a time from log_file. Will raise
        TimeoutError if a line read hangs for more than self.line_wait_timeout
        seconds"""
        waiting_line = None
        while True:
            print("** step counter: ", step_counter)
            line = log_file.readline()
            if not line:
                if waiting_line is None:
                    waiting_line = std.time.time()
                std.time.sleep(0.1)
                total_waiting = std.time.time() - waiting_line
                if total_waiting >= self.line_wait_timeout:
                    raise TimeoutError(
                        "The provided start pattern {} could not be matched \
                        within the specified time interval of {} seconds".format(
                            self.pattern, self.line_wait_timeout
                        )
                    )
            else:
                # reset counter
                waiting_line = None
            yield line
```

Totally up to you though, as both versions do the same. 👍

**Review by nicoddemus — APPROVED** [review]
Great work!

File: xprocess.py (hunk 1 above)
Comment (northernSage) [reply xprocess.py:217]: As the code was before it would just return `None` if you try fetching the `args` property when it's not implemented/defined. No exceptions would be raised and execution would proceed normaly. I prefer following "Explicit is better than implicit" and if someone tries to fetch and use a property that should be implemented but is not, raise a comprehensive exception and interrupt execution.

File: xprocess.py (hunk 2 above)
Comment (northernSage) [reply xprocess.py:245]: Yep, thanks.

File: xprocess.py (hunk 3 above)
Comment (northernSage) [reply xprocess.py:241]: Hm, yeah. I think I would also like your suggest approach more. I'll give this a second try...

File: xprocess.py (hunk 1 above)
Comment (northernSage) [reply xprocess.py:219]: To illustrate:
```python
import abc

class Parent:
    @abc.abstractproperty
    def myprop(self):
        "my test property"

class MyClass(Parent):
    pass

myclass = MyClass()

print(f"My prop value: {myclass.myprop}")
```
> My prop value: None

Opposed to:
```python
import abc

class Parent:
    @abc.abstractproperty
    def myprop(self):
        "my test property"
        raise NotImplementedError

class MyClass(Parent):
    pass

myclass = MyClass()

print(f"My prop value: {myclass.myprop}")
```

> Traceback (most recent call last):
>   File "test.py", line 15, in <module>
>     print(f"My prop value: {myclass.myprop}")
>   File "test.py", line 8, in myprop
>     raise NotImplementedError
> NotImplementedError

**Comment by northernSage** [comment]
Also, after merging this I think we'll be read for 0.16.0 🎉

File: xprocess.py (hunk 1 above)
Comment (nicoddemus) [reply xprocess.py:219]: My question was that `abc.abstractproperty` should raise `NotImplementedError`, but now I see the issue is that the class doesn't subclass from `abc.ABC` (which is responsible for raising the exception really):

```python
from abc import ABC, abstractproperty

class C(ABC):

    @abstractproperty
    def name(self): ...

c = C()
```

In this case, you can't even instantiate `C`:

```
Traceback (most recent call last):
  File "foo.py", line 10, in <module>
    c = C()
TypeError: Can't instantiate abstract class C with abstract methods name
```

So using `@abc.abstractproperty` without subclassing `abc.ABC` seems like an error here.

Comment (northernSage) [reply xprocess.py:219]: Nice catch! I'm considering releasing a patch to fix this, 0.16.1?

Comment (northernSage) [reply xprocess.py:219]: I just saw that @abc.abstractproperty has been

> Deprecated since version 3.3: It is now possible to use property, property.getter(), property.setter() and property.deleter() with abstractmethod(), making this decorator redundant

so on second thought, I think we could instead update the code and not use `@abc.abstractproperty` at all.

Comment (nicoddemus) [reply xprocess.py:219]: >  I think we could instead update the code and not use @abc.abstractproperty at all.

Agreed! 

Btw, when I said "... seems like an error here" I didn't mean that the code _as it stands now_ was incorrect, just that whoever added `@abstractproperty` incorrectly assumed it would raise `NotImplementedError`, which was not the case until you fixed it by raising it explicitly.

In summary, I agree that no new release is necessary, we should just remove the decorator for sanity's sake (and also because it is deprecated as you mentioned).

Comment (northernSage) [reply xprocess.py:219]: Okay, I'll write a PR addressing this in a few days. 👍🏻
